### PR TITLE
switch state cache to use ref statedata objects to limit memory usage

### PR DESCRIPTION
**Pull request by tersec**
More progress towards https://github.com/status-im/nim-beacon-chain/issues/979

490MB RSS after 115 slots of mainnet.

700MB RSS after 200 slots of mainnet.

There's still some memory leak, but it's slower.

### CODE REVIEW

File: beacon_chain/block_pool.nim [hunk 1]
```diff
@@ -379,42 +380,42 @@ proc putState(pool: BlockPool, state: HashedBeaconState, blck: BlockRef) =
...
+  # block's StateData of odd-numbered epoch in bucket 1, whilst evens
+  # land in bucket 0 (which is handed back to GC in if branch). There
+  # still is a possibility of combinatorial explosion, but this only,
+  # by a constant-factor, worsens things. TODO the actual solution's,
+  # eventually, to switch to CoW and/or ref objects for state and the
+  # hash_tree_root processing.
+  let key = (a: blck.root, b: state.data.slot)
+  if key notin pool.cachedStates[bucketParity]:
+    # Avoid constructing StateData if not necessary
+    var stateData = new StateData
```
Comment (arnetheduck) [inline beacon_chain/block_pool.nim:415]: `(ref StateData)(data: sate, blck: blck)`

Comment (tersec) [reply beacon_chain/block_pool.nim:415]: Oh, much nicer.

Comment (tersec) [reply beacon_chain/block_pool.nim]: https://github.com/status-im/nim-beacon-chain/pull/1007/commits/a9c4a474e011ded46bd4fb159859019dae4ad804

**Comment by stefantalpalaru** [comment]
It doesn't seem to stop growing, after an hour:

img

**Comment by stefantalpalaru** [comment]
I pushed a "devel" merge by mistake. It should go away when merging the PR.

Could it be that old tables are not garbage collected at all?

**Comment by tersec** [comment]
> I pushed a "devel" merge by mistake. It should go away when merging the PR.
> 
> Could it be that old tables are not garbage collected at all?
> 
> Another interesting behaviour is that RSS became much closer to the GC heap size.

Yes, it seems to be growing at about one `BeaconState`/`StateData` per slot. It's as if it's just not being GC'd at all.

I don't see where the stray/remaining references keeping it alive might be coming from; the two places it's read are:

https://github.com/status-im/nim-beacon-chain/blob/13aa5ec3e7a9e0c0edeca440fe9355d74a146b1d/beacon_chain/block_pool.nim#L725-L728

and

https://github.com/status-im/nim-beacon-chain/blob/13aa5ec3e7a9e0c0edeca440fe9355d74a146b1d/beacon_chain/block_pool.nim#L804-L809

Both of which use `=` (which indeed, shows up as `genericAssignAux` to copy the dereferenced `StateData` value into a non-`ref` `var StateData`.

**Comment by stefantalpalaru** [comment]
Could the table's tuple key and the object ref value share some pointers?

**Comment by stefantalpalaru** [comment]
I think StateData objects create cycles between them that the GC can't break. Too many levels to track in BlockRef?

**Comment by tersec** [comment]
> I think StateData objects create cycles between them that the GC can't break. Too many levels to track in BlockRef?

That did look problematic, and I switched out `StateData` for `HashedBeaconState`:
```
diff --git a/beacon_chain/beacon_node_types.nim b/beacon_chain/beacon_node_types.nim
index eae79b1..5c845d2 100644
--- a/beacon_chain/beacon_node_types.nim
+++ b/beacon_chain/beacon_node_types.nim
@@ -134,7 +134,7 @@ type
     db*: BeaconChainDB
 
     cachedStates*:
-      array[2, Table[tuple[a: Eth2Digest, b: Slot], ref StateData]] ##\
+      array[2, Table[tuple[a: Eth2Digest, b: Slot], ref HashedBeaconState]] ##\
     ## Dual BeaconChainDBs operates as a pool allocator which handles epoch
     ## boundaries which don't align with an ongoing latency of availability
     ## of precalculated BeaconStates from the recent past.
diff --git a/beacon_chain/block_pool.nim b/beacon_chain/block_pool.nim
index 3ff6b0c..9d10c8c 100644
--- a/beacon_chain/block_pool.nim
+++ b/beacon_chain/block_pool.nim
@@ -235,8 +235,10 @@ proc init*(T: type BlockPool, db: BeaconChainDB,
     pending: initTable[Eth2Digest, SignedBeaconBlock](),
     missing: initTable[Eth2Digest, MissingBlock](),
     cachedStates: [
-      initTable[tuple[a: Eth2Digest, b: Slot], ref StateData](initialSize = 2),
-      initTable[tuple[a: Eth2Digest, b: Slot], ref StateData](initialSize = 2)
+      initTable[tuple[a: Eth2Digest, b: Slot],
+        ref HashedBeaconState](initialSize = 2),
+      initTable[tuple[a: Eth2Digest, b: Slot],
+        ref HashedBeaconState](initialSize = 2)
     ],
     blocks: blocks,
     tail: tailRef,
@@ -383,7 +385,8 @@ proc putState(pool: BlockPool, state: HashedBeaconState, blck: BlockRef) =
   if state.data.slot mod BUCKET_SLOT_LENGTH == 0:
     # The lookback window's >= BUCKET_SLOT_LENGTH and <= BUCKET_SLOT_LENGTH*2.
     pool.cachedStates[bucketParity] =
-      initTable[tuple[a: Eth2Digest, b: Slot], ref StateData](initialSize = 2)
+      initTable[tuple[a: Eth2Digest, b: Slot],
+        ref HashedBeaconState](initialSize = 2)
 
   # Need to be able to efficiently access states for both attestation
   # aggregation and to process block proposals going back to the last
@@ -411,9 +414,10 @@ proc putState(pool: BlockPool, state: HashedBeaconState, blck: BlockRef) =
   # hash_tree_root processing.
   let key = (a: blck.root, b: state.data.slot)
   if key notin pool.cachedStates[bucketParity]:
-    # Avoid constructing StateData if not necessary
-    pool.cachedStates[bucketParity][key] =
-      (ref StateData)(data: state, blck: blck)
+    var foo:ref HashedBeaconState
+    foo = new HashedBeaconState
+    foo[] = state
+    pool.cachedStates[bucketParity][key] = foo
 
 proc add*(
     pool: var BlockPool, blockRoot: Eth2Digest,
@@ -723,12 +727,11 @@ proc rewindState(pool: BlockPool, state: var StateData, bs: BlockSlot):
       let key = (a: parBs.blck.root, b: parBs.slot)
 
       try:
-        state = cachedState[key][]
+        state.data = cachedState[key][]
       except KeyError:
         continue
       let ancestor = ancestors.pop()
-      when false:
-        doAssert state.blck == ancestor.refs
+      state.blck = ancestor.refs
 
       beacon_state_data_cache_hits.inc()
       trace "Replaying state transitions via in-memory cache",
@@ -802,7 +805,8 @@ proc getStateDataCached(pool: BlockPool, state: var StateData, bs: BlockSlot): b
 
   for poolStateCache in pool.cachedStates:
     try:
-      state = poolStateCache[(a: bs.blck.root, b: bs.slot)][]
+      state.data = poolStateCache[(a: bs.blck.root, b: bs.slot)][]
+      state.blck = bs.blck
       beacon_state_data_cache_hits.inc()
       return true
     except KeyError:
```

And it's still eating memory (this time, using `eth2_network_simulation`, since it cycles through mainnet slots twice as fast as testnet1) by what appears to be the same mechanism as this PR.

**Comment by arnetheduck** [comment]
`seq` + reuse instance is looking increasingly attractive, no? :)

**Comment by tersec** [comment]
> `seq` + reuse instance is looking increasingly attractive, no? :)

Yeah, using `seq` next, at least. I avoided `seq` because I wasn't using `ref`, so it seemed prudent to keep as many things in-place as possible (it's a similar reason I didn't use LRU at first -- typical LRU implementations copy things around, and the things in this case were multiply-layered/composed 2+ MB objects).

**Comment by mratsim** [comment]
We can crate a sharedSeq type

```Nim
type SharedSeq[T] {.shallow.} = object
  buf: seq[T]
```

which will have ref semantics. It's better than ref seq as it avoids double GC alloc.

**Comment by arnetheduck** [comment]
meh, `shallow` is an abomination, it hides a mutable reference to the same thing behind magic, changing the fundamental semantics of `object` and `seq` where it's used. 

If by double gc alloc you mean the additional allocation for the `seq` that holds the references to the cache objects, that's really not our biggest concern since it's done once only (or it can be an array) for the lifetime of the application.

**Comment by stefantalpalaru** [comment]
I don't understand the seq+reusal proposal. My theory is that as long as we have multiple objects in memory, objects sharing pointers in a certain pattern that the GC's cycle breaker can't handle, we are bound to leak memory.

Looking at HashedBeaconState, its "data" field is a BeaconState which holds multiple sequences of Eth2Digest, PendingAttestation, Eth1Data, etc.

In the generated C code, sequence members appear as struct pointers. Could we be sharing sequences between objects?

**Comment by tersec** [comment]
> I don't understand the seq+reusal proposal. My theory is that as long as we have multiple objects in memory, objects sharing pointers in a certain pattern that the GC's cycle breaker can't handle, we are bound to leak memory.
> 
> Looking at HashedBeaconState, its "data" field is a BeaconState which holds multiple sequences of Eth2Digest, PendingAttestation, Eth1Data, etc.
> 
> In the generated C code, sequence members appear as struct pointers. Could we be sharing sequences between objects?

SSZ objects are almost always trees, and at most DAGs.

**Comment by tersec** [comment]
https://github.com/status-im/nim-beacon-chain/pull/1007/commits/bded8d77289fe952dd103ac9a78db9543975af6a has been sitting stably at 285-300MB RSS through mainnet slot 115, finalizing through epoch 17, on `make eth2_network_simulation`.

**Comment by arnetheduck** [comment]
> I don't understand the seq+reusal proposal. 

```
type BlockPool = object
  states: seq[ref StateData]

template withState(blockslot, stuff) =
  closest_state = for state in states:
    find_closest_state()
  remove(states, closest_state)
  rewind(closest_state, blockslot)
  stuff()
  add(states, closest_state)

withState(someslot):
  state_transition(closest_rewound_state)
```

notably, there's no copy - the cache owns the state and makes sure no two things use it at once. `states` would grow to a limit (which coincides with the supported nesting)

**Comment by tersec** [comment]
> > I don't understand the seq+reusal proposal.
> 
> ```
> type BlockPool = object
>   states: seq[ref StateData]

After @stefantalpalaru 's observation about the `BlockRef` in `StateData` and whatever was going on in GC, I'd probably not immediately revert to using `StateData`, though it does have advantages. If it works, okay, but there's a lot to be said for not having potential `ref`s between objects in the cache hanging around.

**Comment by stefantalpalaru** [comment]
Mem leak avoided:

img

**Pull request merged**